File: apps/mobile/lib/features/qr/data/validation_queue_repository.dart
import 'package:drift/drift.dart';
import 'package:uuid/uuid.dart';

import '../../../core/database/app_database.dart';

/// File d'attente locale des validations d'étape ([PendingValidations]) —
/// alimentée par [QrFlowNotifier] quand une tentative de validation échoue
/// pour une raison réseau (pas de connexion), rejouée une par une par
/// `SyncService` via [ValidationRepository] (le même chemin d'upload qu'en
/// ligne — aucune duplication de logique multipart).
class ValidationQueueRepository {
  ValidationQueueRepository({required AppDatabase database}) : _db = database;

  final AppDatabase _db;
  final _uuid = const Uuid();

  /// Met une validation en attente et renvoie le `clientEventId` généré,
  /// afin qu'un futur envoi (immédiat ou différé) soit reconnu comme le
  /// même événement côté serveur.
  Future<String> enqueue({
    required String missionStepId,
    required String qrToken,
    required double latitude,
    required double longitude,
    double? accuracy,
    required bool isMocked,
    required DateTime recordedAt,
    required String photoPath,
  }) async {
    final clientEventId = _uuid.v4();
    await _db.into(_db.pendingValidations).insert(
          PendingValidationsCompanion.insert(
            clientEventId: clientEventId,
            missionStepId: missionStepId,
            qrToken: qrToken,
            latitude: latitude,
            longitude: longitude,
            accuracy: Value(accuracy),
            isMocked: Value(isMocked),
            recordedAt: recordedAt,
            photoPath: photoPath,
          ),
        );
    return clientEventId;
  }

  /// Nombre de validations encore `pending`/`failed` — utilisé par la
  /// pastille de connectivité et l'écran de synchronisation.
  Stream<int> watchPendingCount() {
    final query = _db.selectOnly(_db.pendingValidations)
      ..addColumns([_db.pendingValidations.id.count()])
      ..where(
        _db.pendingValidations.syncStatus.equalsValue(SyncStatus.pending) |
            _db.pendingValidations.syncStatus.equalsValue(SyncStatus.failed),
      );
    return query
        .map((row) => row.read(_db.pendingValidations.id.count()) ?? 0)
        .watchSingle();
  }

  Stream<int> watchFailedCount() {
    final query = _db.selectOnly(_db.pendingValidations)
      ..addColumns([_db.pendingValidations.id.count()])
      ..where(_db.pendingValidations.syncStatus.equalsValue(SyncStatus.failed));
    return query
        .map((row) => row.read(_db.pendingValidations.id.count()) ?? 0)
        .watchSingle();
  }

  /// Validations prêtes à être envoyées — une par une (upload multipart),
  /// même règles de backoff/forçage que [GpsQueueRepository.nextBatch].
  Future<List<PendingValidation>> nextBatch({
    required int limit,
    required bool force,
    required int maxAutoRetries,
  }) {
    final now = DateTime.now();
    final query = _db.select(_db.pendingValidations)
      ..where((t) {
        var predicate = t.syncStatus.equalsValue(SyncStatus.pending) |
            t.syncStatus.equalsValue(SyncStatus.failed);
        if (!force) {
          predicate = predicate & t.retryCount.isSmallerThanValue(maxAutoRetries);
          predicate = predicate &
              (t.nextRetryAt.isNull() | t.nextRetryAt.isSmallerOrEqualValue(now));
        }
        return predicate;
      })
      ..orderBy([(t) => OrderingTerm.asc(t.recordedAt)])
      ..limit(limit);
    return query.get();
  }

  Future<void> markUploading(int id) {
    return (_db.update(_db.pendingValidations)..where((t) => t.id.equals(id)))
        .write(
      const PendingValidationsCompanion(
        syncStatus: Value(SyncStatus.uploading),
      ),
    );
  }

  Future<void> markSynced(int id) {
    return (_db.update(_db.pendingValidations)..where((t) => t.id.equals(id)))
        .write(
      const PendingValidationsCompanion(
        syncStatus: Value(SyncStatus.synced),
        lastError: Value(null),
        nextRetryAt: Value(null),
      ),
    );
  }

  Future<void> markFailed({
    required int id,
    required int previousRetryCount,
    String? error,
  }) {
    final retryCount = previousRetryCount + 1;
    return (_db.update(_db.pendingValidations)..where((t) => t.id.equals(id)))
        .write(
      PendingValidationsCompanion(
        syncStatus: const Value(SyncStatus.failed),
        retryCount: Value(retryCount),
        lastError: Value(error),
        nextRetryAt: Value(_nextRetryAt(retryCount)),
      ),
    );
  }

  Future<void> resetStaleUploading() {
    return (_db.update(_db.pendingValidations)
          ..where((t) => t.syncStatus.equalsValue(SyncStatus.uploading)))
        .write(
      const PendingValidationsCompanion(syncStatus: Value(SyncStatus.pending)),
    );
  }
}

/// Backoff exponentiel partagé avec [GpsQueueRepository] : 5s, 10s, 20s,
/// 40s... plafonné à 5 minutes.
DateTime _nextRetryAt(int retryCount) {
  final seconds = (5 * (1 << retryCount.clamp(0, 10))).clamp(5, 300);
  return DateTime.now().add(Duration(seconds: seconds));
}

File: apps/mobile/src/database/validation-queue.repository.ts
import { getDatabase } from './db';
import { PendingValidationRow } from '../types/sync.types';
import { StepValidationPayload } from '../types/tracking.types';

export interface EnqueueValidationParams {
  clientEventId?: string;
  missionStepId: string;
  qrToken: string;
  latitude: number;
  longitude: number;
  accuracy?: number | null;
  isMocked?: boolean;
  recordedAt?: string;
  photoPath: string;
}

export const ValidationQueueRepository = {
  enqueue(params: EnqueueValidationParams): string {
    const db = getDatabase();
    const clientEventId = params.clientEventId || `val_${Date.now()}_${Math.random().toString(36).substring(2, 9)}`;
    const recordedAt = params.recordedAt || new Date().toISOString();
    const createdAtDevice = new Date().toISOString();

    db.runSync(
      `INSERT OR IGNORE INTO pending_validations (
        client_event_id, mission_step_id, qr_token, latitude, longitude,
        accuracy, is_mocked, recorded_at, photo_path, created_at_device,
        sync_status, retry_count
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 'pending', 0)`,
      [
        clientEventId,
        params.missionStepId,
        params.qrToken,
        params.latitude,
        params.longitude,
        params.accuracy ?? null,
        params.isMocked ? 1 : 0,
        recordedAt,
        params.photoPath,
        createdAtDevice,
      ]
    );

    return clientEventId;
  },

  getNextBatch(limit = 5, force = false, maxAutoRetries = 8): PendingValidationRow[] {
    const db = getDatabase();
    const nowIso = new Date().toISOString();

    if (force) {
      return db.getAllSync<PendingValidationRow>(
        `SELECT * FROM pending_validations 
         WHERE sync_status != 'synced' AND sync_status != 'uploading'
         ORDER BY recorded_at ASC LIMIT ?`,
        [limit]
      );
    }

    return db.getAllSync<PendingValidationRow>(
      `SELECT * FROM pending_validations 
       WHERE sync_status IN ('pending', 'failed')
         AND retry_count < ?
         AND (next_retry_at IS NULL OR next_retry_at <= ?)
       ORDER BY recorded_at ASC LIMIT ?`,
      [maxAutoRetries, nowIso, limit]
    );
  },

  markUploading(id: number): void {
    const db = getDatabase();
    db.runSync(`UPDATE pending_validations SET sync_status = 'uploading' WHERE id = ?`, [id]);
  },

  markSynced(id: number): void {
    const db = getDatabase();
    db.runSync(`DELETE FROM pending_validations WHERE id = ?`, [id]);
  },

  markFailed(id: number, previousRetryCount: number, error: string): void {
    const db = getDatabase();
    const nextRetryCount = previousRetryCount + 1;
    const delaySeconds = Math.min(300, 5 * Math.pow(2, Math.max(0, nextRetryCount - 1)));
    const nextRetryAt = new Date(Date.now() + delaySeconds * 1000).toISOString();

    db.runSync(
      `UPDATE pending_validations 
       SET sync_status = 'failed', retry_count = ?, last_error = ?, next_retry_at = ?
       WHERE id = ?`,
      [nextRetryCount, error, nextRetryAt, id]
    );
  },

  resetStaleUploading(): void {
    const db = getDatabase();
    db.runSync(`UPDATE pending_validations SET sync_status = 'pending' WHERE sync_status = 'uploading'`);
  },

  getPendingCount(): number {
    const db = getDatabase();
    const row = db.getFirstSync<{ count: number }>(
      `SELECT COUNT(*) as count FROM pending_validations WHERE sync_status != 'synced'`
    );
    return row?.count || 0;
  },

  getFailedCount(): number {
    const db = getDatabase();
    const row = db.getFirstSync<{ count: number }>(
      `SELECT COUNT(*) as count FROM pending_validations WHERE sync_status = 'failed'`
    );
    return row?.count || 0;
  },

  toPayload(row: PendingValidationRow): StepValidationPayload {
    return {
      clientEventId: row.client_event_id,
      qrToken: row.qr_token,
      latitude: row.latitude,
      longitude: row.longitude,
      accuracy: row.accuracy ?? 0,
      isMocked: row.is_mocked === 1,
      recordedAt: row.recorded_at,
    };
  },
};
